### ISSUE TITLE
Scalac 2.11.7+ / java.lang.AssertionError: assertion failed: Can't convert form ...

### ISSUE THREAD

**Issue by reza-samei** [scala/bug#10815]
I get this error on compile time!
```sbt
[error] (rpc/compile:compileIncremental) java.lang.AssertionError: assertion failed: Can't convert from REF(trait Future) to REF(class AsyncResult) in unit CliLauncher.scala at source-~/projectfolder/module/rpc/src/main/scala/.../rpc/CliLauncher.scala,line-29,offset=775
```
The code: 
```scala
def loadConfig: AsyncResult[SimpleError, InternalConfig] = for {
    configFile <- args.required[String]("config").leftMap(_.asInstanceOf[SimpleError])
    config = ConfigFactory.parseFile(new File(configFile))
    internalConfig <- InternalConfig.extract(config)
} yield internalConfig
```
and `AsyncResult`:
```scala
final class AsyncResult[+L, +R](
    val underlay: Future[StdEither[L, R]]
) extends AnyVal {
    def asFuture: Future[StdEither[L, R]] = underlay
}

object AsyncResult {
   // implicit class to impliment map, flatMap, and ...
}
```
complete error log:
```sbt
java.lang.AssertionError: assertion failed: Can't convert from REF(trait Future) to REF(class AsyncResult) in unit CliLauncher.scala at source-/home/reza/workspace/nathan/module/rpc/src/main/scala/com/bisphone/nathan/rpc/CliLauncher.scala,line-29,offset=775
        at scala.tools.nsc.backend.icode.GenICode$ICodePhase.adapt(GenICode.scala:1091)
        at scala.tools.nsc.backend.icode.GenICode$ICodePhase.scala$tools$nsc$backend$icode$GenICode$ICodePhase$$genLoad(GenICode.scala:1028)
        at scala.tools.nsc.backend.icode.GenICode$ICodePhase$$anonfun$14.apply(GenICode.scala:1123)
        at scala.tools.nsc.backend.icode.GenICode$ICodePhase$$anonfun$14.apply(GenICode.scala:1120)
        at scala.collection.immutable.List.map(List.scala:273)
        at scala.tools.nsc.backend.icode.GenICode$ICodePhase.genLoadLabelArguments(GenICode.scala:1120)
        at scala.tools.nsc.backend.icode.GenICode$ICodePhase.genLoadApply6$1(GenICode.scala:759)
        at scala.tools.nsc.backend.icode.GenICode$ICodePhase.scala$tools$nsc$backend$icode$GenICode$ICodePhase$$genLoad(GenICode.scala:809)
        at scala.tools.nsc.backend.icode.GenICode$ICodePhase.scala$tools$nsc$backend$icode$GenICode$ICodePhase$$genLoad(GenICode.scala:923)
        at scala.tools.nsc.backend.icode.GenICode$ICodePhase.genLoadIf(GenICode.scala:373)
        at scala.tools.nsc.backend.icode.GenICode$ICodePhase.scala$tools$nsc$backend$icode$GenICode$ICodePhase$$genLoad(GenICode.scala:536)
        at scala.tools.nsc.backend.icode.GenICode$ICodePhase.genLoadLabelDef$1(GenICode.scala:505)
        at scala.tools.nsc.backend.icode.GenICode$ICodePhase.scala$tools$nsc$backend$icode$GenICode$ICodePhase$$genLoad(GenICode.scala:507)
        at scala.tools.nsc.backend.icode.GenICode$ICodePhase.scala$tools$nsc$backend$icode$GenICode$ICodePhase$$genStat(GenICode.scala:181)
        at scala.tools.nsc.backend.icode.GenICode$ICodePhase$$anonfun$genStat$1.apply(GenICode.scala:155)
        at scala.tools.nsc.backend.icode.GenICode$ICodePhase$$anonfun$genStat$1.apply(GenICode.scala:155)
        at scala.collection.LinearSeqOptimized$class.foldLeft(LinearSeqOptimized.scala:124)
        at scala.collection.immutable.List.foldLeft(List.scala:84)
        at scala.tools.nsc.backend.icode.GenICode$ICodePhase.scala$tools$nsc$backend$icode$GenICode$ICodePhase$$genLoad(GenICode.scala:922)
        at scala.tools.nsc.backend.icode.GenICode$ICodePhase.gen(GenICode.scala:123)
        at scala.tools.nsc.backend.icode.GenICode$ICodePhase.gen(GenICode.scala:71)
        at scala.tools.nsc.backend.icode.GenICode$ICodePhase.gen(GenICode.scala:148)
        at scala.tools.nsc.backend.icode.GenICode$ICodePhase.gen(GenICode.scala:98)
        at scala.tools.nsc.backend.icode.GenICode$ICodePhase.gen(GenICode.scala:71)
        at scala.tools.nsc.backend.icode.GenICode$ICodePhase.gen(GenICode.scala:89)
        at scala.tools.nsc.backend.icode.GenICode$ICodePhase.gen(GenICode.scala:67)
        at scala.tools.nsc.backend.icode.GenICode$ICodePhase.apply(GenICode.scala:63)
        at scala.tools.nsc.Global$GlobalPhase$$anonfun$applyPhase$1.apply$mcV$sp(Global.scala:440)
        at scala.tools.nsc.Global$GlobalPhase.withCurrentUnit(Global.scala:431)
        at scala.tools.nsc.Global$GlobalPhase.applyPhase(Global.scala:440)
        at scala.tools.nsc.Global$GlobalPhase$$anonfun$run$1.apply(Global.scala:398)
        at scala.tools.nsc.Global$GlobalPhase$$anonfun$run$1.apply(Global.scala:398)
        at scala.collection.Iterator$class.foreach(Iterator.scala:742)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1194)
        at scala.tools.nsc.Global$GlobalPhase.run(Global.scala:398)
        at scala.tools.nsc.backend.icode.GenICode$ICodePhase.run(GenICode.scala:55)
        at scala.tools.nsc.Global$Run.compileUnitsInternal(Global.scala:1501)
        at scala.tools.nsc.Global$Run.compileUnits(Global.scala:1486)
        at scala.tools.nsc.Global$Run.compileSources(Global.scala:1481)
        at scala.tools.nsc.Global$Run.compile(Global.scala:1582)
        at xsbt.CachedCompiler0.run(CompilerInterface.scala:116)
        at xsbt.CachedCompiler0.run(CompilerInterface.scala:95)
        at xsbt.CompilerInterface.run(CompilerInterface.scala:26)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at sbt.compiler.AnalyzingCompiler.call(AnalyzingCompiler.scala:107)
        at sbt.compiler.AnalyzingCompiler.compile(AnalyzingCompiler.scala:53)
        at sbt.compiler.AnalyzingCompiler.compile(AnalyzingCompiler.scala:47)
        at sbt.compiler.MixedAnalyzingCompiler$$anonfun$compileScala$1$1.apply$mcV$sp(MixedAnalyzingCompiler.scala:50)
        at sbt.compiler.MixedAnalyzingCompiler$$anonfun$compileScala$1$1.apply(MixedAnalyzingCompiler.scala:50)
        at sbt.compiler.MixedAnalyzingCompiler$$anonfun$compileScala$1$1.apply(MixedAnalyzingCompiler.scala:50)
        at sbt.compiler.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:74)
        at sbt.compiler.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:49)
        at sbt.compiler.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:64)
        at sbt.compiler.IC$$anonfun$compileInternal$1.apply(IncrementalCompiler.scala:160)
        at sbt.compiler.IC$$anonfun$compileInternal$1.apply(IncrementalCompiler.scala:160)
        at sbt.inc.IncrementalCompile$$anonfun$doCompile$1.apply(Compile.scala:66)
        at sbt.inc.IncrementalCompile$$anonfun$doCompile$1.apply(Compile.scala:64)
        at sbt.inc.IncrementalCommon.cycle(IncrementalCommon.scala:32)
        at sbt.inc.Incremental$$anonfun$1.apply(Incremental.scala:72)
        at sbt.inc.Incremental$$anonfun$1.apply(Incremental.scala:71)
        at sbt.inc.Incremental$.manageClassfiles(Incremental.scala:99)
        at sbt.inc.Incremental$.compile(Incremental.scala:71)
        at sbt.inc.IncrementalCompile$.apply(Compile.scala:54)
        at sbt.compiler.IC$.compileInternal(IncrementalCompiler.scala:160)
        at sbt.compiler.IC$.incrementalCompile(IncrementalCompiler.scala:138)
        at sbt.Compiler$.compile(Compiler.scala:155)
        at sbt.Compiler$.compile(Compiler.scala:141)
        at sbt.Defaults$.sbt$Defaults$$compileIncrementalTaskImpl(Defaults.scala:913)
        at sbt.Defaults$$anonfun$compileIncrementalTask$1.apply(Defaults.scala:904)
        at sbt.Defaults$$anonfun$compileIncrementalTask$1.apply(Defaults.scala:902)
        at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
        at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
        at sbt.std.Transform$$anon$4.work(System.scala:63)
        at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
        at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
        at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
        at sbt.Execute.work(Execute.scala:237)
        at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
        at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
        at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
        at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
[error] (nathan-rpc/compile:compileIncremental) java.lang.AssertionError: assertion failed: Can't convert from REF(trait Future) to REF(class AsyncResult) in unit CliLauncher.scala at source-/home/reza/workspace/nathan/module/rpc/src/main/scala/com/bisphone/nathan/rpc/CliLauncher.scala,line-29,offset=775
```

Note: It will compile with `2.11.5` & `2.11.6` versions; but `2.11.7` and newer version will fail.

**Comment by reza-samei** [comment]
It's fine with 2.12(.5);  but I had a hardtime to cross-compile all legacy libs in work.
But steel I need to compile part of it in 2.11.x because Flink don't support 2.12 yet!
When I try to run the 2.11.6 compiled version, I get a runtime exception (that I don't get in 2.12 compiled version):
```
java.lang.ClassCastException: scala.concurrent.impl.Promise$DefaultPromise cannot be cast to AsyncResultjava.lang.ClassCastException: scala.concurrent.impl.Promise$DefaultPromise cannot be cast to AsyncResult
```
It seems that 2.11 is not going to get a fix for this, but FYI it seems that it related to for-comprehension.

And please accept my apologize to being late at responding.

**Comment by SethTisue** [comment]
The compiler back end (which is where the stack trace shows the exception to be coming from) was completely replaced between Scala 2.11 and 2.12, so there's no possibility of backporting a fix from 2.12 to 2.11.

I'm going to close the ticket since the problem doesn't exist in 2.12. On 2.11, I'm afraid you'll need to try and find some kind of workaround.
